pkg/apply/prune/prune.go

```diff
@@ -153,6 +153,17 @@ func (p *Pruner) Prune(
 					}
 				}
 
+				// Remove the object from inventory if it was determined that the object should not be pruned,
+				// because it had recently been applied. This probably means that the object is in the inventory
+				// more than one time with a different group (e.g. kind Ingress and apiGroups networking.k8s.io & extensions)
```

Comment:
Are you sure that `ApplyPreventedDeletionError` **always** means that there's a "duplicate" object in the source of truth? I'm concerned there may be side effects for other use cases that we're not testing for yet.

While I did re-write this check into the new Filter pattern, I'm not entirely sure what use case it was meant to handle in the first place.

Would it help if we tracked not just the UID but also the GVK that was applied, so it doesn't overlap when the API group is different?

Reply:
> Are you sure that ApplyPreventedDeletionError always means that there's a "duplicate" object in the source of truth? I'm concerned there may be side effects for other use cases that we're not testing for yet.

I am not 100% sure, but can't come up with any other possible reasons for it!

> While I did re-write this check into the new Filter pattern, I'm not entirely sure what use case it was meant to handle in the first place.

This check originates from https://github.com/kubernetes-sigs/cli-utils/pull/113. Perhaps it'd be valuable to hear @seans3 thoughts on this change?

> Would it help if we tracked not just the UID but also the GVK that was applied, so it doesn't overlap when the API group is different?

I am not sure tracking UID+GVK is the best option as then we'll have two references to exactly the same object. E.g. in the apply set we'd apply `Ingress.networking.k8s.io` and then we'd prune `Ingress.extensions` right away as it's not filtered by `CurrentUIDFilter`. In reality it's the same object and the filter was created to prevent pruning objects just applied.

Reply:
Where there are duplicate object references, one of them should be removed from the inventory. A duplicate reference is when multiple `ObjMetadata` (Group/Kind/Namespace/Name) point to the same object (by `UID`). The `UID` check exists specifically for this case (it also exists in `kubectl apply/prune`). The `UID` is the only metadata field which can definitively determine the uniqueness of an object.

As previously described, the most common instance this duplicate reference happens is when a `Deployment` or an `Ingress` is updated to the latest version from the `extensions` group and then applied. The `ObjMetatdata` check is unable to determine the objects are the same, but the `UID` check figures it out (and the old `ObjMetadata` reference should be removed).

I will LGTM and approve when we are all on the same page.

Reply:
Gotcha. Sounds like something we need an end to end test for to make sure we don’t break it in future refactors. Maybe just a v1beta1 to v1 upgrade, for example. 

Right now, the uid filter skips the delete, but doesnt remove the old object from the inventory. 

I suspect there’s also a better place to put this code. The inv_set_task has code to decide what stays in the inventory using information from the inventory cache in the context. You don’t need to check error types there. 

In the future, I want to make the inventory update dynamically, but for now all that logic is in the inv_set_task.

Reply:
On second thought, inv_set_task doesn’t know the difference between the types of skips. So maybe this way is fine.

Reply:
@manfredlift can you add a uid filter test under test/e2e ?

we have namespace filter, skip invalid, and dependency filter e2e tests, but none for upgrading to a new resource version.

Reply:
@karlkfi upgrading to a new version would not actually result in this issue as in the inventory we track objects using groupKind (without the version) **not** the GVK.

To demonstrate the issue, I added an e2e test that upgrades an `Event` (co-existing/co-habitating in both `v1` and `events.k8s.io/v1`) from `v1` to `events.k8s.io/v1`. This example is a bit contrived, but in k8s v1.25, which the e2e tests use, I could not any better examples (in the past the same scenario has existed for `Ingress` and `Deployment` though!)

Reply:
@seans3 added an e2e test to demonstrate & verify the issue.

Reply:
Would it make more sense to use 2 CRDs then, so it doesn’t break when k8s is upgraded?

Reply:
I had the same thoughts initially, but...

AFAIK, CRDs don't support being cohabitated (https://github.com/kubernetes/kubectl/issues/186#issuecomment-369612606) where the object with different API groups is stored in storage only once (with same UID) and converted into either group on the fly. For CRDs, this support is only there between versions of the same API group.

These are some examples of cohabitating resources:
https://github.com/kubernetes/kubernetes/blob/8fdaac238e4f7f560066ea3a324cdbcaae564ac9/pkg/kubeapiserver/default_storage_factory_builder.go#L122-L129

Reply:
Cool. That makes it more clear what we have this exception for and when we can retire it, assuming k8s doesn't add new ones in the future.

Can you add this link to that list in the filter description or e2e test comment?
https://github.com/kubernetes/kubernetes/blob/v1.26.0/pkg/kubeapiserver/default_storage_factory_builder.go#L124

Reply:
Done!